phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java

```diff
@@ -0,0 +1,422 @@
+        }
+    }
+
+    private boolean isCellDeleted(List<Cell> deleteMarkers, CompactionRowVersion rowVersion,
+            Cell cell, boolean storeLevel) {
+        int i = 0;
+        for (Cell dm : deleteMarkers) {
+            if ((storeLevel ||
+                    Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(),
+                    cell.getFamilyLength(),
+                    dm.getFamilyArray(), dm.getFamilyOffset(), dm.getFamilyLength()) == 0) &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(),
+                            cell.getQualifierLength(),
+                            dm.getQualifierArray(), dm.getQualifierOffset(),
+                            dm.getQualifierLength()) == 0) {
+                if (dm.getType() == Cell.Type.Delete) {
+                    deleteMarkers.remove(i);
+                    if (rowVersion.columnDeleteMarkers == null) {
+                        rowVersion.columnDeleteMarkers = new ArrayList<>();
+                    }
+                    rowVersion.columnDeleteMarkers.add(dm);
+                }
+                return true;
+            }
+            i++;
+        }
+        return false;
+    }
+
+    private long getNextRowVersionTimestamp(List<List<Cell>> columns) {
+        long ts = 0;
+        for (List<Cell> column : columns) {
+            Cell firstCell = column.get(0);
+            if (firstCell.getType() == Cell.Type.DeleteFamily ||
+                    firstCell.getType() == Cell.Type.DeleteFamilyVersion) {
+                break;
+            }
+            if (firstCell.getType() == Cell.Type.DeleteColumn ||
+                    firstCell.getType() == Cell.Type.Delete) {
+                continue;
+            }
+            if (ts < firstCell.getTimestamp()) {
+                ts = firstCell.getTimestamp();
+            }
+        }
+        return ts;
+    }
+
+    private boolean formRowVersions(List<List<Cell>> columns,
+                                    List<CompactionRowVersion> rowVersions,
+                                    boolean storeLevel) {
+        Cell lastDeleteFamilyMarker = null;
+        Cell lastDeleteFamilyVersionMarker = null;
+        List<Cell> columnDeleteMarkers = null;
+        int version = 0;
+        while (!columns.isEmpty()) {
+            long ts = getNextRowVersionTimestamp(columns);
+            CompactionRowVersion rowVersion = null;
+            // Form the next row version by picking the first cell from each column if the cell
+            // is not masked by a delete marker
+            for (List<Cell> column : columns) {
+                Cell firstCell = column.remove(0);
+                if (firstCell.getType() == Cell.Type.DeleteFamily ||
+                        firstCell.getType() == Cell.Type.DeleteFamilyVersion) {
+                    if (firstCell.getType() == Cell.Type.DeleteFamily) {
+                        if (firstCell.getTimestamp() >= ttlWindowStart &&
```

Comment:
Don't we want to preserve the delete marker if it is inside the maxLookback window even if KEEP_DELETED_CELLS == FALSE?

Reply:
Yes, we need to preserve it. The decision is done later in prepareResults. This is just an optimization to early terminate the row version generation here. However, there is a typo in the if statement. It should read if (firstCell.getTimestamp() < maxLookbackWindowStart &&. Good catch!